### PR TITLE
Add support for VoodooInput

### PR DESCRIPTION
**Pull request by kprinssu**
This PR will remove the MT2 simulator and actuator in favour of VoodooInput.

Why VoodooInput? The VoodooI2C team (@alexandred @ben9923 @blankmac and myself) and @vit9696 and @usr-sse2 have had lengthy discussions about how to manage the VoodooI2C MT2 simulator and VoodooPS2's fork. We eventually came to the conclusion that it would be best to refactor the MT2 simulator into a different project, called VoodooInput. 

I originally created a stand-alone VoodooI2C engine to integrate with VoodooInput called VoodooInputEngine (https://github.com/kprinssu/VoodooInputEngine), but this has quickly become out of date. 

My suggestion is that we integrate the VoodooInputEngine directly into VoodooInput and "tightly" couple VoodooInput via git submodules. Assuming that the VoodooInput headers do not change, this will ensure that that VoodooI2C will always be up to date with changes in VoodooInput.

@Goshin brought up a very good point in https://github.com/alexandred/VoodooI2C/issues/244#issuecomment-589636208. Well, we should address this and similar issues at the VoodooInput level. My suggestion is that VoodooI2C should be agnostic of what kind of device we are emulating (multi-touch or simple mouse) and let VoodooInput handle it.

### CODE REVIEW

**Comment by kprinssu** [comment]
This PR depends on this PR for VoodooInput, https://github.com/acidanthera/VoodooInput/pull/9

**Comment by kprinssu** [comment]
@Goshin can you give this branch ago, you'll need to grab the latest version of VoodooInput (built from source) and let me know if there are any regressions?

**Comment by Goshin** [comment]
Should clean up the `message.transducers` before sending a new message to VoodooInput.

https://github.com/kprinssu/VoodooI2C/blob/f9778dc6b8d30eb83c551a0ac7696e2a7e34a768/Multitouch%20Support/Native/VoodooI2CNativeEngine.cpp#L14-L31

Otherwise, VoodooInput will get one additional old finger data in the loop because:

https://github.com/acidanthera/VoodooInput/blob/4e471ae559312c2d0ec4efabc6f5596f3b594486/VoodooInput/VoodooInputSimulator/VoodooInputSimulatorDevice.cpp#L85

This can cause the single-finger touch not responding issue after a multi-finger touch has been once ever processed. the bug is easy to replicate.

File: Multitouch Support/Native/VoodooI2CNativeEngine.cpp [hunk 1]
```diff
@@ -12,72 +12,92 @@
...
+        VoodooInputTransducer* inputTransducer = &message.transducers[i];
+        
+        if (!transducer) {
+            continue;
+        }
+        
+        inputTransducer->id = transducer->id;
+        inputTransducer->secondaryId = transducer->secondary_id;
+        
+        inputTransducer->type = VoodooInputTransducerType::FINGER;
```
Comment (Goshin) [inline Multitouch Support/Native/VoodooI2CNativeEngine.cpp]: Why override the transducer type?

**Comment by kprinssu** [comment]
Good points, I will clean it up later tonight.

File: Multitouch Support/Native/VoodooI2CNativeEngine.cpp [hunk 2]
```diff
@@ -12,72 +12,92 @@
...
-
-    // Initialize simulator device
-    if (!simulator->init(NULL) || !simulator->attach(this))
-        goto exit;
-    else if (!simulator->start(this)) {
-        simulator->detach(this);
-        goto exit;
+    
+    parentProvider = NULL;
+    parentProvider = OSDynamicCast(VoodooI2CMultitouchInterface, provider);
```
Comment (ben9923) [inline Multitouch Support/Native/VoodooI2CNativeEngine.cpp]: Doesn't `OSDynamicCast` return `nullptr` on failure?
```suggestion
    parentProvider = OSDynamicCast(VoodooI2CMultitouchInterface, provider);
```

**Comment by kprinssu** [comment]
@Goshin give the new set of changes a go and let me know if there any issues.

File: Multitouch Support/Native/VoodooI2CNativeEngine.cpp [hunk 3]
```diff
@@ -22,6 +22,7 @@ MultitouchReturn VoodooI2CNativeEngine::handleInterruptReport(VoodooI2CMultitouc
     for (int i = 0; i < event.contact_count; i++) {
         VoodooI2CDigitiserTransducer* transducer = (VoodooI2CDigitiserTransducer*) event.transducers->getObject(i);
         VoodooInputTransducer* inputTransducer = &message.transducers[i];
+        memset(inputTransducer, 0, sizeof(VoodooInputTransducer));
```
Comment (Goshin) [inline Multitouch Support/Native/VoodooI2CNativeEngine.cpp]: Well I mean clean up the `message.transducers` before the loop. 😂
like this:
```diff
MultitouchReturn VoodooI2CNativeEngine::handleInterruptReport(VoodooI2CMultitouchEvent event, AbsoluteTime timestamp) {
    if (!voodooInputInstance) {
        return MultitouchReturnContinue;
    }

+   message = {};
    message.timestamp = timestamp;
    message.contact_count = event.contact_count;
    
    for (int i = 0; i < event.contact_count; i++) {
```
or
```diff
MultitouchReturn VoodooI2CNativeEngine::handleInterruptReport(VoodooI2CMultitouchEvent event, AbsoluteTime timestamp) {
    if (!voodooInputInstance) {
        return MultitouchReturnContinue;
    }

    message.timestamp = timestamp;
    message.contact_count = event.contact_count;
+   memset(message.transducers, 0, VOODOO_INPUT_MAX_TRANSDUCERS * sizeof(VoodooInputTransducer));
    
    for (int i = 0; i < event.contact_count; i++) {
```

**Pull request merged**